cmd/server/cadence/cadence.go

```diff
@@ -111,34 +114,72 @@ func BuildCLI(releaseVersion string, gitRevision string) *cli.App {
 				},
 			},
 			Action: func(c *cli.Context) error {
-				fxApp := fx.New(
-					config.Module,
-					logfx.Module,
-					dynamicconfigfx.Module,
-					fx.Provide(func() appContext {
-						return appContext{
-							CfgContext: config.Context{
-								Environment: getEnvironment(c),
-								Zone:        getZone(c),
-							},
-							ConfigDir: getConfigDir(c),
-							RootDir:   getRootDir(c),
-							Services:  getServices(c),
-						}
-					}),
-					Module,
-				)
+				stoppedWg := &sync.WaitGroup{}
 
-				ctx := context.Background()
-				if err := fxApp.Start(ctx); err != nil {
-					return err
-				}
+				ctx, cancel := context.WithCancel(context.Background())
+				defer cancel()
 
-				// Block until FX receives a shutdown signal
-				<-fxApp.Done()
+				services := getServices(c)
+				// non blocking channel that can accommodate at least 3 errors per service (worst case).
+				errChan := make(chan error, len(services)*3)
+				for _, serv := range getServices(c) {
+					stoppedWg.Add(1)
+					go func(s string) {
+						defer stoppedWg.Done()
+						fxApp := fx.New(
+							fx.Module(s,
+								config.Module,
+								dynamicconfigfx.Module,
+								logfx.Module,
+								metricsfx.Module,
+								clockfx.Module,
+								fx.Provide(
+									func() appContext {
+										return appContext{
+											CfgContext: config.Context{
+												Environment: getEnvironment(c),
+												Zone:        getZone(c),
+											},
+											ConfigDir: getConfigDir(c),
+											RootDir:   getRootDir(c),
+										}
+									},
+								),
+								Module(s),
+							),
+						)
 
-				// Stop the application
-				return fxApp.Stop(ctx)
+						//  If any of the start hooks return an error, Start short-circuits, calls Stop, and returns the inciting error.
+						if err := fxApp.Start(ctx); err != nil {
+							errChan <- fmt.Errorf("service %s start: %w", s, err)
+							return
+						}
+
+						select {
+						// Block until FX receives a shutdown signal
+						case <-fxApp.Done():
+						}
+
+						// Stop the application
+						err := fxApp.Stop(ctx)
+						if err != nil {
+							errChan <- fmt.Errorf("service %s stop: %w", s, err)
+						}
+					}(serv)
+				}
```

Comment:
I think we should move this to it's own function to reduce the level of indentation